mlir/include/mlir/Dialect/MIOpen/XdlopsCodeSelection.h

```diff
@@ -467,221 +441,62 @@ struct XdlopsCodeSelection {
-      num_threads_blk = 16;
-      num_output_blks = 1;
-      num_regs_xdlops = num_regs_blk * num_output_blks;
-      m = 16;
-      n = 16;
-      k = 8;
-      cycles = 32;
-      k_base = 2;
-    } else if (instr == amdgpu::MFMAInstr::f32_32x32x2f32) {
-      group_size = 4;
-      num_groups_blk = 1;
-      num_regs_blk = group_size * num_groups_blk;
-      num_threads_blk = 16;
-      num_output_blks = 4;
-      num_regs_xdlops = num_regs_blk * num_output_blks;
-      m = 16;
-      n = 16;
-      k = 2;
-      cycles = 32;
-      k_base = 2;
-    } else if (instr == amdgpu::MFMAInstr::f32_4x4x2bf16) {
-      group_size = 4;
-      num_groups_blk = 1;
-      num_regs_blk = group_size * num_groups_blk;
-      num_threads_blk = 64;
-      num_output_blks = 1;
-      num_regs_xdlops = 4;
-      m = 4;
-      n = 64;
-      k = 2;
-      cycles = 8;
-      k_base = 2;
-    } else if (instr == amdgpu::MFMAInstr::i32_32x32x8i8) {
-      group_size = 4;
-      num_groups_blk = 4;
-      num_regs_blk = group_size * num_groups_blk;
-      num_threads_blk = 32;
-      num_output_blks = 1;
-      num_regs_xdlops = num_regs_blk * num_output_blks;
-      m = 32;
-      n = 32;
-      k = 8;
-      cycles = 64;
-      k_base = 4;
-    } else if (instr == amdgpu::MFMAInstr::i32_16x16x16i8) {
-      group_size = 4;
-      num_groups_blk = 1;
-      num_regs_blk = group_size * num_groups_blk;
-      num_threads_blk = 16;
-      num_output_blks = 1;
-      num_regs_xdlops = num_regs_blk * num_output_blks;
-      m = 16;
-      n = 16;
-      k = 16;
-      cycles = 32;
-      k_base = 4;
-    } else {
-      llvm_unreachable("Unsupported case as mfmaInstr not selected!\n");
-    }
+    // Rename this field in future refactoring
+    int64_t vectorNumber = imms.size();
+    int64_t numXdlops = vectorNumber / (MRepeats * NRepeats);
+    int64_t mCandidate = MPerXdlops / numXdlops;
+    int64_t nCandidate = NPerXdlops / numXdlops;
+    int64_t mfmaOutDim = std::min(mCandidate, nCandidate);
+    int64_t m = mfmaOutDim, n = mfmaOutDim;
+
+    int64_t kPerInputArg =
+        math_util::integer_divide_ceil(waveSize, mfmaOutDim * blocksMfma);
+    // k_base is the length of the step you must take in the k dimension of
+    // your input when preparing the next argument on a thread.
+    int64_t k_base = k / kPerInputArg;
+
+    int64_t nInputs = (m * blocksMfma * k) / waveSize;
+    Type argType =
+        nInputs == 1 ? dataType : VectorType::get({nInputs}, dataType);
+
+    int64_t nOutputs = (m * n * blocksMfma) / waveSize;
+    Type vectorElem;
+    if (dataType.isa<IntegerType>())
+      vectorElem = b.getI32Type();
+    else
+      vectorElem = b.getF32Type();
+    VectorType vectorType = VectorType::get({nOutputs}, vectorElem);
+
+    constexpr int64_t group_size = 4;
+    int64_t rowsPerOutput = std::min(waveSize / n, m / group_size);
+    int64_t blocksPerOutput =
+        math_util::integer_divide_ceil(waveSize, rowsPerOutput * n);
```

Comment:
I start to get lost form here on again...

Could I ask you to draw a 1:1 reference mapping according to just one sample of xdlops?

Something like `16x16x4f16` with  below would work:

![image-20220121163541641](https://user-images.githubusercontent.com/3454501/184951572-cc0bb4bd-1a72-4af2-86ea-8aa31c1d1aab.png)